docs/proposals/20210628-workloadspread.md

```diff
@@ -0,0 +1,223 @@
+  - `creatingPods`: the map of Pods that just allocated to this subset by webhook
+    > creatingPods contains information about pods whose creation was processed by
+      the webhook handler but not yet been observed by the WorkloadSpread controller.
+      A pod will be in this map from the time when the webhook handler processed the
+      creation request to the time when the pod is seen by controller. \
+      The key in the map is the name of the pod and the value is the time when the webhook
+      handler process the creation request. If the real creation didn't happen and a pod is
+      still in this map, it will be removed from the list automatically by WorkloadSpread controller
+      after some time.\
+      If everything goes smooth this map should be empty for the most of the time.
+      Large number of entries in the map may indicate problems with pod creations.
+  - `deletingPods`: the map of Pods been deleted or evicted that just noticed by webhook
+    > deletingPods is similar with creatingPods and it contains information about pod deletion.
+  - `subsetUnscheduledStatus`: contains the details for the unscheduled subset.
+     - `unschedulable`: unschedulable is true indicates this Subset cannot be scheduled. The default is false.
+     - `unscheduledTime`: last time for unscheduled.
+     - `failedCount`: the number of subset was marked with unschedulable.
+
+### Requirements
+
+WorkloadSpread requires Pod webhook to inject rules into Pods and notice the deletion and eviction of Pods.
+
+So if the `PodWebhook` feature-gate is set to `false`, WorkloadSpread will also be disabled.
+
+### Implementation Details/Notes/Constraints
+
+WorkloadSpread have both webhook and controller. Controller should collaborate with webhook to maintain WorkloadSpread status together. \
+The controller is responsible for calculating the real status, and the webhook mainly counts missingReplicas and records the creation or deletion entry of Pod into map.
+
+#### Replicas control
+
+##### Pod creation 
+When a Pod is creating and its workload has a related WorkloadSpread, webhook will check the subsetStatuses in the WorkloadSpread one by one:
+
+1. if `missingReplicas` > 0 or `missingReplicas` = -1 and this subset is not unschedulable, choose this subset temporarily \
+   (1). update `missingReplicas` -= 1 and put this pod into `creatingPods` map and try to update subsetStatus. \
+   (2). if it conflicts, get the WorkloadSpread again and go back to step (1). \
+   (3). if it succeeds, inject the rules of this subset into the Pod. 
+2. if `missingReplicas` = 0, go to the next subset
+3. if there is no available subset left, just let the Pod go
+
+##### Pod deletion 
+Also, when a Pod that belongs to a subset is deleting or evicting, webhook will put it into the `deletingPods` map in subsetStatus and update `missingReplicas` += 1.
+
+##### Update Pod Status
```

Comment:
Shouldn't a Pod be removed from the `creatingPod` map when its status is changed to "running"?

Reply:
Controller will remove Pod entry that has been created from `creatingPods` map. The definition of `create` is that the controller discovers the Pod, regardless of whether its status is running, pending or otherwise.

Reply:
In addition, the reason we update missingReplicas and remove entry from creatingMap is that the kubelet can update Pod phase to `failed`  in a very short time due to unsatisfied resources. Controller can't update the missingReplicas in time, which will cause error for Pods spread in multiple subset. Therefore，webhook may be better in this scenario.